main.ml

```
@@ -568,29 +568,47 @@ let process root_dir name =
+          let uuid = List.Assoc.find probe_result.Xapi_storage.Volume.Types.configuration ~equal:String.equal "uuid" in
+          let open Deferred.Or_error in
+          match probe_result.Xapi_storage.Volume.Types.sr, probe_result.Xapi_storage.Volume.Types.complete, uuid with
+          | Some sr_stat, false, Some _uuid ->
+              return (`Attachable (probe_result.configuration, {
+                  Storage_interface.name_label = sr_stat.Xapi_storage.Volume.Types.name;
+                  name_description = sr_stat.Xapi_storage.Volume.Types.description;
+                  total_space = sr_stat.Xapi_storage.Volume.Types.total_space;
+                  free_space = sr_stat.Xapi_storage.Volume.Types.free_space;
+                  clustered = sr_stat.Xapi_storage.Volume.Types.clustered;
+                  health = match sr_stat.Xapi_storage.Volume.Types.health with
+                  | Xapi_storage.Volume.Types.Healthy _ -> Healthy
+                  | Xapi_storage.Volume.Types.Recovering _ -> Recovering
+                }));
+          | Some _sr, true, _ ->
+              errorf "A configuration cannot be suitable for both SR.create and SR.attach: %a" pp_probe_result probe_result
+          | Some _sr, _, None ->
+              errorf "A configuration is not attachable without a uuid: %a" pp_probe_result probe_result
+          | None, false, Some _uuid ->
+              errorf "A configuration with a uuid is cannot be incomplete: %a" pp_probe_result probe_result
+          | None, false, None ->
+              return (`Incomplete probe_result.configuration)
+          | None, true, Some _uuid ->
+              return (`Creatable probe_result.configuration)
+          | None, true, None ->
+              errorf "A configuration cannot be complete without a uuid: %a" pp_probe_result probe_result)
+        |> Deferred.Or_error.combine_errors
+        |> Deferred.Result.map_error ~f:(fun err -> err |> Error.to_string_hum |> R.rpc_of_string)
```

So this is where we return an RPC error is any of the probe results are inconsistent?

---

Yes, I wanted to return all the errors, to make debugging/development easier, instead of stopping at the first one.
The error might be a bit too verbose though, but if the plugin is working correctly a user should never see such an error.

---

Ok, that sounds good. So if any of the SR probe results is incorrect, the whole `SR.probe` call will fail with a combined error? That sounds good to me, the plugins should work properly.

---

Yes, I wouldn't want to hide errors in the plugins, we should fix them as soon as we find them.